Sort ordering of sources is preserved

ntl opened the pull request:
Resolves #1609

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

ntl commented on lib/rubygems/source.rb:
```diff
@@ -67,7 +67,9 @@ def <=>(other)
-      @uri.to_s <=> other.uri.to_s
+      return 1 unless @uri.to_s == other.uri.to_s
+
+      0
```
The idea here is that we still want to treat the sources as equal (e.g. return `0`) when the URLs are the same, but we want the first source to have priority over the second if they differ. This causes the original ordering of the sources to be preserved.

drbrain replied:
Can you move this ↑ into a comment above line 70?

It'll make it harder to accidentally revert.

ntl commented:
I discovered a bug with this PR that I believe I have fixed. However, I am a bit lost as to where to put the tests. I need to test the resolver's `search_for` method, as I have changed how it sorts. Specifically, I have to group the possible specs by their source, and _then_ sort each group. This ensures that the source ordering is respected by the resolver.

segiddins commented on lib/rubygems/resolver.rb:
```diff
@@ -233,7 +233,13 @@ def search_for(dependency)
+    possibles.
+      group_by(&:source).
+      sort.
+      flat_map { |_, s_list|
```
`flat_map` isn't available on 1.8.7

ntl replied:
Thanks! Corrected.

segiddins commented:
@ntl maybe in one of the resolver spec files?

ntl commented:
I appreciate the guidance, @segiddins . I added a test case to prove that the sources are grouped before versions are sorted. I had to add enough different specs to cause ruby's `#sort` to reveal the flaw of my earlier implementation.

ntl commented:
Discovered that I had an additional 1.8.7 compatibility issue. Fixed now.

segiddins commented on test/rubygems/test_gem_resolver.rb:
```diff
@@ -683,6 +683,36 @@ def test_second_level_backout
+
+    resolver = Gem::Resolver.new [dependency], set
+
+    assert_resolves_to [spec_A_1], resolver
```
wouldn't this pass even on master, because '1.0.0' is already the highest version?

ntl replied:
That is true. And, actually, after further testing, I have discovered that rubygems resorts the sources alphabetically before resolving dependencies. I am continuing to investigate this further.

ntl commented:
@segiddins After setting up a more controlled environment (multiple gem servers on my development laptop with exactly one gem, `foo`, at different versions), I have discovered a much more succinct way of surfacing the original issue. The test has been updated.

ntl commented on lib/rubygems/resolver.rb:
```diff
@@ -233,8 +233,29 @@ def search_for(dependency)
+
+        activation_requests << activation_request
+      end
+    end
```
Unfortunately, the source ordering has to be taken into account before sorting the versions and platform locality. The original sort ordering supplied by the user is not available in this lexical context. We can only rely on the assumption that the gems were returned from `#find_possible` in order of their source. This assumption holds because the fetcher iterates over each source in order to retrieve the specs.

ntl commented:
A final note, I apologize for taking so long to get back to this. I went on vacation and then got a bit behind after I got back.

segiddins commented:
👍 from me

ntl commented:
@segiddins what is the process for getting this PR accepted? Is there any further work needed from me, or any way I can otherwise assist the process?

AaronBetterOfficeApps commented:
:+ Any timeline on this being merged?

sbellware commented:
👍

segiddins commented:
I was just waiting for another contributor to review the changes

indirect commented:
@homu r+

homu commented:
:pushpin: Commit b0bf979 has been approved by `indirect`

<!-- @homu r=indirect b0bf979e2a4427792c84008971c11a83052e35e6 -->

homu commented:
:hourglass: Testing commit b0bf979 with merge aa663c8...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/rubygems/rubygems/builds/155312194)

ntl commented:
Thanks so much! This will help me out a lot.

sbellware commented:
Hi Guys! Is there any expected timeline for release of this fix? Thanks!
